### PR TITLE
refs #1172: Add short introduction to promises

### LINKED ISSUE #1172: Small introduction to promises
This is very similar to #1119, there is no real documentation in `Mojo::Promise`. I think it would be nice to explain at the least the basics of how to use promises, their possible states, what returning promises in the `->then(...)` callbacks means and so on.

### PR DESCRIPTION
**Pull request by lammel**
### Summary

This PR aims to improve the documentation of Mojo::Promise with a short introduction into the subject. Only documentation is added, no code changes.

The `println` in the examples should be replaced with something more useful, but should be ok for discussion for now.

### Motivation

Improve docs for beginners and newcomers that may find promises and their composability useful.

### References

Should close #1172

### CODE REVIEW

**Comment by lammel** [comment]
To much Go lately ;-)

I actually wanted to show how the get could be implemented by using a
promise, to have a useful example of asynchronous operation that takes some
time to complete.
Maybe I should add that get_p does what the example shows.

2018-04-13 23:43 GMT+02:00 Dan Book <notifications@github.com>:

File: lib/Mojo/Promise.pm [hunk 1]
```diff
@@ -182,7 +182,63 @@ Mojo::Promise - Promises/A+
...
+
+For example a new promise can be created to resolve only after both 
+GET operations have completed using the method L<Mojo::Promise/"all">.
+
+  my $promise1 = get("http://mojolicious.org");
+  my $promise2 = get("https://perlbrew.pl/");
+  my $both = Mojo::Promise->all($promise1, $promise2)->then(sub {
+    my ($value1, $value2) = @_;
+    println "Mojolicous website returns " . $value1;
+    println "Perlbrew website returns " . $valuu2;
```
Comment (marcusramberg) [inline lib/Mojo/Promise.pm]: s/valuu2/value2/ in addition to say instead of println as mentioned

**Comment by marcusramberg** [comment]
I think that our basic example should not not reimplement get_p.

**Comment by lammel** [comment]
Thanks for the feedback!

The basic example was taken from the synopsis, but maybe is not the best
way based on your feedback. I'd like to have something people will
associated with a longer running task (web request, brew beer, background
job,  long calculation) and then do the all example to actually put the
promise code to use.

Ideas welcome, on what a basic example could do...

2018-04-14 10:45 GMT+02:00 Marcus Ramberg <notifications@github.com>:

**Comment by kraih** [comment]
Looks like this pull request went stale.

**Comment by lammel** [comment]
Sorry, yes I ran out of time to spare.
But I'm still supporting this PR, but will need some more time (approx 3
weeks)

BR Roland

2018-04-27 10:11 GMT+02:00 Sebastian Riedel <notifications@github.com>:

**Comment by kraih** [comment]
Understood, please open a new pull request when you are ready.

**Pull request closed without merging**